### ISSUE TITLE
Consider supporting nan for use with other node runtimes

### ISSUE THREAD

**Issue by grofit** [mathjax/MathJax-node#102]
I am just prototyping and was trying to use MathJax-node from within Electron but it fails to run, and if I try to convert it using the electron rebuilder it blows up, so I was wondering if there was any chance you could support nan so it would work across all node runtimes?

https://www.npmjs.com/package/nan

**Comment by pkra** [comment]
Thanks for the posting. From a quick glance, nan is about native addons so I'm not sure how this applies to MathJax-node (except for its dependencies). Is this related to the problems with jsdom mentioned in #101?

**Comment by grofit** [comment]
No no, not related to that issue. NaN is more about letting a module work in NodeJS and IO.js and any other Node derivative without any additional work.

I have side stepped the issue by just including the entire mathjax project locally (removing some output types I wont need), it just felt it would have been more succinct using the mathjax node version but like I say it refuses to load in Electrons version of NodeJS (which uses a newer version of the V8 runtime).

Feel free to close if you want.

**Comment by pkra** [comment]
> NaN is more about letting a module work in NodeJS and IO.js and any other Node derivative without any additional work.

Hm. We thought it did (cf. #95). But #101 reported issues with a) later versions of jsdom b) CSP related issues.

> I have side stepped the issue by just including the entire mathjax project locally (removing some output types I wont need), it just felt it would have been more succinct using the mathjax node version but like I say it refuses to load in Electrons version of NodeJS (which uses a newer version of the V8 runtime).

I'm not sure I follow. You're including MathJax but not MathJax-node, correct? This seems to indicate you are using MathJax not MathJax-node, i.e., you're using MathJax to render in a browser context and not in a Nodejs context (where MathJax does not work in itself). Is that correct?

**Comment by grofit** [comment]
lol sorry to confuse.

I *WAS* using `MathJax-node` this way I could just pass back some latex to the nodejs back end, have mathjax node process it directly and return the SVG markup then I could do what I wanted with that.

However as this didnt include correctly I went back to trying to use the normal non node mathjax via the browser, but I couldnt include it via cdn I needed it locally so I didnt want to have to install the whole 18mb or whatever it was.

**Comment by pkra** [comment]
> I WAS using MathJax-node this way I could just pass back some latex to the nodejs back end, have mathjax node process it directly and return the SVG markup then I could do what I wanted with that.

Got it. Could you provide us with more information on the issues you've encountered with MathJax-node in Electron?

**Comment by grofit** [comment]
Sure but its not as helpful as I am sure you were expecting... the issue I got was the same as this guy:

https://github.com/atom/electron/issues/1774

Where it basically says the module did not self register, however ultimately that then leads you to here:

https://github.com/atom/electron/blob/master/docs/tutorial/using-native-node-modules.md

Which is where I noticed some modules on NPM have the keyword `nan` and after doing a bit of digging and trying their electron rebuild (which failed) it just seemed like the module would need updating for that version of nodejs runtime.

Again this is probably a REALLY niche scenario so I wouldn't lose any sleep over it.

**Comment by pkra** [comment]
> Which is where I noticed some modules on NPM have the keyword nan and after doing a bit of digging and trying their electron rebuild (which failed) it just seemed like the module would need updating for that version of nodejs runtime.

I thought "native add-on" refers to add-ons that are not written in JavaScript but C/C++, cf https://iojs.org/api/addons.html (nb: it's listed in the toc as "C/C++ addons"). So I don't understand how any of this refers to MathJax-node (beyond, of course, dependencies like jsdom).

> Again this is probably a REALLY niche scenario so I wouldn't lose any sleep over it.

No worries there. Just trying to understand if this is primarily an upstream issue with our dependencies.

**Comment by grofit** [comment]
Oh you are probably right, I thought native addon meant it depended upon something only within the nodejs runtime, i.e `require("fs");` would not exist within the browser.

**Comment by pkra** [comment]
Ok. Tentively marking this `upstream`. FWIW, Electron is a use case we want to support.

**Comment by mangecoeur** [comment]
FYI, I ran into a similar problem early this summer and I suspect you might have had the same issue. The "native" conflict was caused by the jsdom dependency which in the `master` branch is set to 3.x, which isn't compatible with Electron (on `develop` this has been changed). The latest versions of jsdom should be compatible and also simplify the internals wrt native dependencies (that's the `nan` thing), which should remove a lot of worries around cross compatibility. With a bit of luck, using the develop branch of mathjax-node with the latest jsdom should work with Electron.
